File: w3ggo/gamedata.go
package w3ggo

import "encoding/binary"

type leaveGameBlock struct {
	playerID uint8
	reason   string
	result   string
}

type commandBlock struct {
	playerID uint8
	actions  []Action
}

type timeslotBlock struct {
	timeIncrement uint16
	commandBlocks []commandBlock
}

type chatMessageBlock struct {
	playerID uint8
	mode     uint32
	message  string
}

type gameDataBlockType int

const (
	gdLeaveGame gameDataBlockType = iota
	gdTimeslot
	gdChatMessage
)

type gameDataBlock struct {
	typ       gameDataBlockType
	leaveGame leaveGameBlock
	timeslot  timeslotBlock
	chat      chatMessageBlock
}

func parseGameData(data []byte, isPost202 bool) []gameDataBlock {
	var blocks []gameDataBlock
	pos := 0

	for pos < len(data) {
		if pos >= len(data) {
			break
		}
		id := data[pos]
		pos++

		switch id {
		case 0x17:
			b, ok := parseLeaveGame(data, &pos)
			if ok {
				blocks = append(blocks, gameDataBlock{typ: gdLeaveGame, leaveGame: b})
			}
		case 0x1a, 0x1b, 0x1c:
			if pos+4 > len(data) {
				pos = len(data)
			} else {
				pos += 4
			}
		case 0x1e, 0x1f:
			b, ok := parseTimeslot(data, &pos, isPost202)
			if ok {
				blocks = append(blocks, gameDataBlock{typ: gdTimeslot, timeslot: b})
			}
		case 0x20:
			b, ok := parseChatMessage(data, &pos)
			if ok {
				blocks = append(blocks, gameDataBlock{typ: gdChatMessage, chat: b})
			}
		case 0x22:
			if pos >= len(data) {
				break
			}
			ln := int(data[pos])
			pos++
			if pos+ln > len(data) {
				pos = len(data)
			} else {
				pos += ln
			}
		case 0x23:
			if pos+10 > len(data) {
				pos = len(data)
			} else {
				pos += 10
			}
		case 0x2f:
			if pos+8 > len(data) {
				pos = len(data)
			} else {
				pos += 8
			}
		default:
			// unknown block id - skip and continue (mirrors w3gjs behavior)
			continue
		}
	}
	return blocks
}

func parseLeaveGame(data []byte, pos *int) (leaveGameBlock, bool) {
	reason := gdReadHex(data, pos, 4)
	if *pos >= len(data) {
		return leaveGameBlock{}, false
	}
	playerID := data[*pos]
	*pos++
	result := gdReadHex(data, pos, 4)
	*pos += 4
	return leaveGameBlock{playerID: playerID, reason: reason, result: result}, true
}

func parseTimeslot(data []byte, pos *int, isPost202 bool) (timeslotBlock, bool) {
	if *pos+2 > len(data) {
		return timeslotBlock{}, false
	}
	byteCount := int(binary.LittleEndian.Uint16(data[*pos:]))
	*pos += 2
	if *pos+2 > len(data) {
		return timeslotBlock{}, false
	}
	timeIncrement := binary.LittleEndian.Uint16(data[*pos:])
	*pos += 2

	actionBlockLastOffset := *pos + byteCount - 2
	if actionBlockLastOffset > len(data) {
		actionBlockLastOffset = len(data)
	}

	var commandBlocks []commandBlock
	for *pos < actionBlockLastOffset && *pos < len(data) {
		if *pos >= len(data) {
			break
		}
		playerID := data[*pos]
		*pos++
		if *pos+2 > len(data) {
			break
		}
		actionBlockLength := int(binary.LittleEndian.Uint16(data[*pos:]))
		*pos += 2
		end := *pos + actionBlockLength
		if end > len(data) {
			end = len(data)
		}
		actionData := data[*pos:end]
		actions := parseActions(actionData, isPost202)
		*pos = end
		commandBlocks = append(commandBlocks, commandBlock{playerID: playerID, actions: actions})
	}

	return timeslotBlock{timeIncrement: timeIncrement, commandBlocks: commandBlocks}, true
}

func parseChatMessage(data []byte, pos *int) (chatMessageBlock, bool) {
	if *pos >= len(data) {
		return chatMessageBlock{}, false
	}
	playerID := data[*pos]
	*pos++
	*pos += 2 // byteCount
	if *pos >= len(data) {
		return chatMessageBlock{}, false
	}
	flags := data[*pos]
	*pos++
	var mode uint32
	if flags == 0x20 {
		if *pos+4 > len(data) {
			return chatMessageBlock{}, false
		}
		mode = binary.LittleEndian.Uint32(data[*pos:])
		*pos += 4
	}
	message := gdReadZTS(data, pos)
	return chatMessageBlock{playerID: playerID, mode: mode, message: message}, true
}

func gdReadHex(data []byte, pos *int, n int) string {
	if *pos+n > len(data) {
		return ""
	}
	s := encodeHex(data[*pos : *pos+n])
	*pos += n
	return s
}

func gdReadZTS(data []byte, pos *int) string {
	start := *pos
	for *pos < len(data) && data[*pos] != 0 {
		*pos++
	}
	s := string(data[start:*pos])
	if *pos < len(data) {
		*pos++
	}
	return s
}

File: w3grs/src/parser/game_data.rs
use super::action::{parse_actions, Action};

#[derive(Debug, Clone)]
pub struct LeaveGameBlock {
    pub player_id: u8,
    pub reason: String,
    pub result: String,
}

#[derive(Debug, Clone)]
pub struct CommandBlock {
    pub player_id: u8,
    pub actions: Vec<Action>,
}

#[derive(Debug, Clone)]
pub struct TimeslotBlock {
    pub time_increment: u16,
    pub command_blocks: Vec<CommandBlock>,
}

#[derive(Debug, Clone)]
pub struct ChatMessageBlock {
    pub player_id: u8,
    pub mode: u32,
    pub message: String,
}

#[derive(Debug, Clone)]
pub enum GameDataBlock {
    LeaveGame(LeaveGameBlock),
    Timeslot(TimeslotBlock),
    ChatMessage(ChatMessageBlock),
}

pub fn parse_game_data(data: &[u8], is_post_202: bool) -> Vec<GameDataBlock> {
    let mut blocks = Vec::new();
    let mut pos = 0usize;

    while pos < data.len() {
        let id = match data.get(pos) {
            Some(&v) => { pos += 1; v }
            None => break,
        };

        match id {
            0x17 => {
                if let Some(b) = parse_leave_game(data, &mut pos) {
                    blocks.push(GameDataBlock::LeaveGame(b));
                }
            }
            0x1a | 0x1b | 0x1c => {
                // skip 4 bytes
                pos = (pos + 4).min(data.len());
            }
            0x1e | 0x1f => {
                if let Some(b) = parse_timeslot(data, &mut pos, is_post_202) {
                    blocks.push(GameDataBlock::Timeslot(b));
                }
            }
            0x20 => {
                if let Some(b) = parse_chat_message(data, &mut pos) {
                    blocks.push(GameDataBlock::ChatMessage(b));
                }
            }
            0x22 => {
                let len = data.get(pos).copied().unwrap_or(0) as usize;
                pos += 1;
                pos = (pos + len).min(data.len());
            }
            0x23 => {
                pos = (pos + 10).min(data.len());
            }
            0x2f => {
                pos = (pos + 8).min(data.len());
            }
            _ => {
                // unknown block id - skip and continue (mirrors w3gjs behavior)
                continue;
            }
        }
    }
    blocks
}

fn parse_leave_game(data: &[u8], pos: &mut usize) -> Option<LeaveGameBlock> {
    let reason = read_hex(data, pos, 4);
    let player_id = ru8(data, pos)?;
    let result = read_hex(data, pos, 4);
    *pos += 4; // skip 4
    Some(LeaveGameBlock { player_id, reason, result })
}

fn parse_timeslot(data: &[u8], pos: &mut usize, is_post_202: bool) -> Option<TimeslotBlock> {
    let byte_count = ru16(data, pos)? as usize;
    let time_increment = ru16(data, pos)?;
    let action_block_last_offset = *pos + byte_count.saturating_sub(2);

    let mut command_blocks = Vec::new();

    while *pos < action_block_last_offset && *pos < data.len() {
        let player_id = ru8(data, pos)?;
        let action_block_length = ru16(data, pos)? as usize;
        let end = (*pos + action_block_length).min(data.len());
        let action_data = &data[*pos..end];
        let actions = parse_actions(action_data, is_post_202);
        *pos = (*pos + action_block_length).min(data.len());
        command_blocks.push(CommandBlock { player_id, actions });
    }

    Some(TimeslotBlock { time_increment, command_blocks })
}

fn parse_chat_message(data: &[u8], pos: &mut usize) -> Option<ChatMessageBlock> {
    let player_id = ru8(data, pos)?;
    *pos += 2; // byteCount
    let flags = ru8(data, pos)?;
    let mut mode = 0u32;
    if flags == 0x20 {
        mode = ru32(data, pos)?;
    }
    let message = read_zts(data, pos);
    Some(ChatMessageBlock { player_id, mode, message })
}

// helpers
fn ru8(data: &[u8], pos: &mut usize) -> Option<u8> {
    let v = *data.get(*pos)?;
    *pos += 1;
    Some(v)
}

fn ru16(data: &[u8], pos: &mut usize) -> Option<u16> {
    if *pos + 2 > data.len() { return None; }
    let v = u16::from_le_bytes([data[*pos], data[*pos+1]]);
    *pos += 2;
    Some(v)
}

fn ru32(data: &[u8], pos: &mut usize) -> Option<u32> {
    if *pos + 4 > data.len() { return None; }
    let v = u32::from_le_bytes([data[*pos], data[*pos+1], data[*pos+2], data[*pos+3]]);
    *pos += 4;
    Some(v)
}

fn read_hex(data: &[u8], pos: &mut usize, n: usize) -> String {
    if *pos + n > data.len() { return String::new(); }
    let s = hex::encode(&data[*pos..*pos+n]);
    *pos += n;
    s
}

fn read_zts(data: &[u8], pos: &mut usize) -> String {
    let start = *pos;
    while *pos < data.len() && data[*pos] != 0 { *pos += 1; }
    let s = String::from_utf8_lossy(&data[start..*pos]).into_owned();
    if *pos < data.len() { *pos += 1; }
    s
}
